usl/usl_kevm.h

```diff
@@ -0,0 +1,241 @@
+struct ommer_t {
+  account_address_t ommer_address;
+  number_t ommer_number;
+};
+using ommer_list_t = std::vector<ommer_t>;
+
+// Block
+using difficulty_t = std::uint64_t;
+using timestamp_t = std::time_t;
+using mix_hash_t = std::uint64_t;
+struct block_t {
+  hash_t previous_hash;
+  hash_t ommers_hash;
+  account_address_t coinbase;
+  hash_t state_root;
+  hash_t transactions_root;
+  hash_t receipts_root;
+  bytes_t logs_bloom;
+  difficulty_t difficulty;
+  number_t number;
+  gas_t gas_limit;
+  gas_t gas_used;
+  timestamp_t timestamp;
+  bytes_t extra_data;
+  mix_hash_t mix_hash;
+  nonce_t block_nonce;
+  ether_t base_fee;
+  hash_t withdrawals_root;
+  ommer_list_t ommer_block_headers;
+};
+
+// Account
+struct account_t {
+  account_address_t account_id;
+  ether_t balance;
+  bytes_t code;
+  nonce_t nonce;
+};
+
+// Access List
+using storage_key_list_t = std::vector<storage_key_t>;
+struct access_pair_t {
+  account_address_t account_id;
+  storage_key_list_t storage_keys;
+};
+using access_list_t = std::vector<access_pair_t>;
+
+// Transaction Type
+enum class tx_type_t { LEGACY, ACCESS_LIST, DYNAMIC_FEE };
+using tx_type_option_t = std::optional<tx_type_t>;
+
+// Message
+using account_address_option_t = std::optional<account_address_t>;
+using sig_v_t = std::uint64_t;
+struct message_t {
+  message_id_t message_id;
+  nonce_t tx_nonce;
+  ether_t tx_gas_price;
+  gas_t tx_gas_limit;
+  account_address_option_t to;
+  ether_t value;
+  sig_v_t sig_v;
+  bytes_t sig_r;
+  bytes_t sig_s;
+  bytes_t data;
+  access_list_t tx_access;
+  chain_id_t tx_chain_id;
+  ether_t tx_priority_fee;
+  ether_t tx_max_fee;
+  tx_type_option_t tx_type;
+};
+
+// Network Callbacks
+using add_account_callback_t = void (*)(account_t);
+using remove_account_callback_t = void (*)(account_address_t);
+
+using get_account_balance_callback_t =
+    std::optional<ether_t> (*)(account_address_t);
+using set_account_balance_callback_t = void (*)(account_address_t, ether_t);
+
+using get_account_code_callback_t =
+    std::optional<bytes_t> (*)(account_address_t);
+using set_account_code_callback_t = void (*)(account_address_t, bytes_t);
+
+using get_account_nonce_callback_t =
+    std::optional<nonce_t> (*)(account_address_t);
+using set_account_nonce_callback_t = void (*)(account_address_t, nonce_t);
+
+using get_account_storage_callback_t =
+    std::optional<storage_value_t> (*)(account_address_t, storage_key_t);
+using set_account_storage_callback_t = void (*)(account_address_t,
+                                                storage_key_t, storage_value_t);
+
+using get_account_orig_storage_callback_t =
+    std::optional<storage_value_t> (*)(account_address_t, storage_key_t);
+using set_account_orig_storage_callback_t = void (*)(account_address_t,
+                                                     storage_key_t,
+                                                     storage_value_t);
+
+using add_message_callback_t = void (*)(message_t);
```

Comment:
I don't entirely understand this. What are you envisioning would call this?

Reply:
So this is something that is not completely clear to me. The KEVM semantics maintains a map that contains messages, specifically a mapping from message ID to message cell. As far as I can tell those message IDs are created by the semantics when a new message cell is created (only place I found is in the driver that reads JSON inputs and generates transactions). Does this mean that this map is internal to the semantics and we do not need to populate/access it through callbacks when the KEVM works for the USL?

I wasn't sure so I added the callbacks and exposed the message ID field, which is not part of the yellow paper specification of a transaction btw.

Reply:
We agreed to remove the message ID and callbacks since they are not applicable in the USL use case.